keps/sig-cli/20190318-kustomize-generators-transformers.md

```diff
@@ -0,0 +1,389 @@
+- Develop a market place of any sort
+  - The ecosystem solutions themselves will have market places.  This allows those ecosystem tools to
+    plug their market places into kubectl via Kustomize.
+
+## Proposal
+
+Introduce 2 new types of `exec` plugins:
+
+- Transformers
+- Generators
+
+#### Terms
+
+- *Virtual Resource*
+  - Resource that is not a server-side Kubernetes API object, but used to configure tools
+  - Examples: kubeconfig, kustomization
+- *Generator*
+  - Kustomize directive that accepts a Virtual Resource as input, and emits non-Virtual Resources as output
+  - Examples: `configMapGenerator`, `secretGenerator`
+- *Transformer*
+  - Kustomize directive that accepts 1 Virtual Resource *and* collection of non-Virtual Resources as input
+    and emits non-Virtual Resources as output.
+  - Examples: `commonAnnotations`, `namespace`
+- *Built-In*
+  - Either a Transformer or Generator that is embedded directly in the `kustomization.yaml` rather than
+    as a separate API.
+
+#### Generators
+
+Generators generate 0 or more Resources for some virtual Resource input.
+
+Generator plugins are run immediately after built-in generators:
+
+- `resources`
+- `bases`
+- `secretGenerator`
+- `configMapGenerator`
+
+and before any built-in transformers such as:
+
+- `namePrefix`
+- `commonAnnotations`
+
+Example: Kustomize secretGenerators and configMapGenerators generate Secrets and ConfigMaps
+from a `kustomize.config.k8s.io/v1beta1/Kustomization` virtual Resource.
+
+Generators have 2 components:
+
+1. Generator virtual Resource API definition
+  - By convention, API has a field `generate`
+1. Generator executable plugin that reads the Resources on STDIN and emits generated Resources on STDOUT.
+
+- A Generator is installed by adding the generator executable to
+  `$XDG_CONFIG_HOME/kustomize/plugins/generators/<API group>`.
+- Generators are added as virtual Resources using the `kustomization.yaml` field `generators`
+
+Steps for each `generator`:
+
+1. Kustomize reads the `generator` entry
+1. Kustomize uses the Resource's *Group* to find the plugin under `$XDG_CONFIG_HOME/kustomize/plugins/generators/`
+1. Kustomize execs the plugin it finds, or exits non-0
+1. Kustomize writes the `generator` Resource to the exec process STDIN
+1. Kustomize reads the set of generated Resources from the exec process STDOUT
+1. Kustomize reads error messages from the exec process STDERR
+1. Kustomize fails if the plugin exits non-0
+1. Kustomize adds the emitted Resources to its set of input Resources (e.g. from `resources`, `bases`, etc).
+
+Notes:
+
+- Kustomize will not execute any plugins for the output Resources at this time.
+- If multiple generators are specified, they are run in the order they are listed.
+
+#### Transforms
+
+Transformers modify existing non-virtual Resources by modifying their fields or replacing them with new Resources.
+
+Transformer plugins are run immediately *before* built-in transformers:
```

Comment:
I don't think we should restrict the transformer plugins to be *before*. They can also be *after* the build-in transformers. Both are useful in certain scenarios. The kustomization.yaml author need to specify whether it is before or after. 

Some *before* transformer plugins I can think about:
- inject a sidecar container to pod template

Some *after* transformer plugins:
- validate the resource configs(the output yaml) according to some specific policy. Since the resource configs are not completely done until performing all the built-in transformers. These validation should be as a post processing step

Reply:
Hm, maybe this should be a separate pattern - `validators`.  I think we want to stick to a small number of highly structured patterns, rather than 1 very flexible pattern.